Update dokka to 1.4

romanowski opened the pull request:
Change the way how we render pages, so we parse .md files and then apply html layouts.

Add basic support for linking.

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/locationProvider.kt:
```diff
@@ -13,10 +13,16 @@ class StaticSiteLocationProviderFactory(private val ctx: DokkaContext) : Locatio
+                // replace title with original markdown file name
+                val original = super.pathTo(node, context)
+                val paths = original.split("/")
+                (paths.dropLast(1) + listOf(node.template.template.name())).joinToString("/")
```
this (`template.template`) naming is a bit strange and confusing, what about naming it `loadedTemplate` and `templateFile` respectively?

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/parser.kt:
```diff
@@ -0,0 +1,357 @@
+import java.net.URL
+
+
+// Thisi is a code taken from dokka, will be removed after PR with fixes is merged into dokka
```
Can we have a link to the pr / pr number here?

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/processors.kt:
```diff
@@ -43,39 +28,27 @@ abstract class BaseStaticSiteProcessor(cxt: DokkaContext) : PageTransformer {
-    protected fun File.asDri(): DRI {
-        val relativePath = root.toPath().relativize(toPath()).toString().replace(File.separatorChar, '.')
-        return DRI("_.$relativePath")
+    data class LoadedTemplete(val template: TemplateFile, val children: List<LoadedTemplete>, val file: File) {
```
`s/LoadedTemplete/LoadedTemplate/` and maybe rename the val `template` to avoid confusion?

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/processors.kt:
```diff
@@ -243,50 +172,83 @@ class SitePagesCreator(cxt: DokkaContext) : BaseStaticSiteProcessor(cxt) {
+            emptySet(),
+            PropertyContainer.empty()
+        )
+        return PartiallyRenderedContend(
```
`s/Contend/Content/`

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/processors.kt:
```diff
@@ -243,50 +172,83 @@ class SitePagesCreator(cxt: DokkaContext) : BaseStaticSiteProcessor(cxt) {
+
+        val driMap = all.flatMap { flatten(it) }.map { it to pathToDri(it) }.toMap()
+
+        fun templeteToPage(myTemplate: LoadedTemplete): StaticPageNode {
```
`s/templeteToPage/templateToPage/`

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/renderer.kt:
```diff
@@ -1,70 +1,108 @@
 
-class ExternalDocsToolRenderer(context: DokkaContext) : org.jetbrains.dokka.base.renderers.html.HtmlRenderer(context) {
+
+// this will be reweitten after PR is merged to dokka
```
is there any link to the PR or its number?

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/processors.kt:
```diff
@@ -43,39 +28,27 @@ abstract class BaseStaticSiteProcessor(cxt: DokkaContext) : PageTransformer {
-
-        override val content: ContentNode = PreRenderedContent(resolved.code, DCI(dri, ContentKind.Empty), mySourceSet)
+        fun title(): String = template.template.title()
+        fun noFrame(): Boolean = template.template.noFrame()
```
It is okay and I am not orthodox about negations but just for everyone's information:
https://schneide.blog/2014/08/03/dont-ever-not-avoid-negative-logic/

So it could be `hasFrame` but `noFrame` works too to me

romanowski replied:
That is a good point.

mkondratek commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 build/
 gradle/
 lib/
+out
```
It is a good practice to add the new lines at the end of file

https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

mkondratek commented on src/main/kotlin/com/virtuslab/dokka/site/templates.kt:
```diff
@@ -80,34 +100,48 @@ data class TemplateFile(val file: File, val rawCode: String, private val setting
+                HtmlRenderer.builder(ctx.markdownOptions).build().render(parser.parse(rendered))
+            }
+            val resources = listSetting("extraCSS") + listSetting("extraJS")
+            layoutTemplate?.let {
```
redundant let call

mkondratek commented:
Btw I saw the following issue (I am not sure whether it's been caused by these changes)

![image](https://user-images.githubusercontent.com/19799111/92486145-0b287c80-f1ec-11ea-81e9-0082d1125ced.png)

index page rendered twice, as the index page and as a subpage

mkondratek commented:
Another thing is that the relative links do not work in real life (comparing to tests) yet. 

Try to reference some `md` file from the other file. The resulting link retains its `md` extension (instead of expected `html`)

romanowski commented:
Relative links does not work, still but we got an issue to fix it #17

romanowski committed:
Update dokka to 1.4

Change the way how we render pages, so we parse .md files and then apply html layouts.

Do no render files other then .md or .html ones

mkondratek reviewed (approved):
lgtm then
